QueuedThreadPool will not reserve a thread with jobs waiting

gregw opened the pull request:
Fix #13187 by:
 + Modified QTP's ReservedThreadExecutor so that it will not reserve a thread if there are queued jobs waiting.
 + Deprecating many meaningless getters that combined the concept of available for execute(task) with available for tryExecute(task).
 + Improved javadoc

gregw committed:
QueuedThreadPool will not reserve a thread with jobs waiting

Fix #13187 by:
 + Modified QTP's ReservedThreadExecutor so that it will not reserve a thread if there are queued jobs waiting.
 + Deprecating many meaningless getters that combined the concept of available for execute(task) with available for tryExecute(task).
 + Improved javadoc

gregw committed:
QueuedThreadPool will not reserve a thread with jobs waiting

Fix #13187 by:
 + Modified QTP's ReservedThreadExecutor so that it will not reserve a thread if there are queued jobs waiting.
 + Deprecating many meaningless getters that combined the concept of available for execute(task) with available for tryExecute(task).
 + Improved javadoc

gregw committed:
QueuedThreadPool will not reserve a thread with jobs waiting

Fix #13187 by:
 + Modified QTP's ReservedThreadExecutor so that it will not reserve a thread if there are queued jobs waiting.
 + Deprecating many meaningless getters that combined the concept of available for execute(task) with available for tryExecute(task).
 + Improved javadoc

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1222,4 +1254,34 @@ private void doRunJob(Runnable job)
+            // If we have more jobs queued than the number of pending reserved threads,
+            // then we have some real jobs in the queue.
+            // So we will not start a reserved thread as it may stop a real job from being executed.
+            if (isTaskWaiting())
```
I am borderline -1 on this change.

This fundamentally changes Jetty's behavior from `EXECUTE_PRODUCE_CONSUME` to `PRODUCE_EXECUTE_CONSUME` when it's under enough load that tasks start getting queued.

In my understanding, that is supposed to help lower max latency at the expense of higher average latency. This sounds reasonable, but how much effect does that really have? And is that a good thing? Wouldn't that have an impact on throughput which could lead to a snowball effect?

Plus, this is fundamentally racy if the number of threads has been tuned to be just enough to handle the load: in such case, spawning a new reserved thread becomes totally random and you still have the issue of a reserved thread idling out while there is a job waiting in the queue depending on the exact timing of this test vs the queuing of the job.

gregw replied:
@lorban I understand your reluctance, so I've moved the javadoc changes to a different PR for 12.1.x and we can take our time on this one.

So here is my thinking:
 + Reserving a thread is always an optional optimisation.  We can run with maxReservedThreads (as it should be called) of 0.
 + When running in a constrained environment of limited threads, then reserving a thread is less important than getting real work done.

Perhaps this should only be conditional on us reaching maxThreads. Prior to that it is better to start a thread to reserve it, than to avoid doing so.  However, I doubt we will have threads queued (at least not for long) if we are at max threads.

lorban replied:
I'm not sure I'm aligned with your view of the importance this optimization could have.

My fear is that a system that needs all its threads to serve a load with the _optimized path_ might not be able to serve that same load with the _unoptimized path_, and would enter a spiral of death by switching to a slower code path when it reaches a certain load.

Maybe I'm over-worrying, but I think I won't be able to find peace of mind without seeing the results of some limit tests showing how QTP+AdaptiveExecutionStrategy behave when they have to handle a load too big for the thread pool for a short duration: do they temporarily degrade performance or do they collapse? Can they recover once the load goes down again or did they go to the sorry place, never coming back?

gregw replied:
+1 for more perf/stress testing.... if only we knew somebody to do that :)

I was interesting that one of the tests that needed to be adjusted was checking how timeouts worked if jobs were queued in the QTP.  The test failed because with this change, we could handle one more request than without it, as instead of being reserved a thread took one more job off the queue to execute.   That felt like a good change to me.

lorban commented:
Definitely, QTP's getters and javadoc need updating so I'm happy with that change.

But the _not reserving a thread if there are queued jobs waiting_ part is IMHO problematic and gives me a feeling of a generic solution trying to solve a special case problem.

My view of the problem is that we have a finite resource (a number of threads) that is shared between different consumers (execution of 'normal' jobs and reservations for 'special' jobs) fighting in a concurrent context. And we're not happy that sometimes it's not the ideal candidate that wins the resource race so we'd like to introduce some extra fairness.

I'm very opinionated on this kind of tweaks as it's never possible to be totally fair without a serious reduction of performance, and unwanted side effects of tweaks that supposedly add some level of fairness on the cheap are hard to foresee.

scscgit commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java:
```diff
@@ -386,4 +388,14 @@ public String toString()
     }
+
+    /**
+     * Is there a better task to do than reserved thread a thread?
```
This sentence seems to have a typo

scscgit commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -611,11 +624,11 @@ public int getMaxReservedThreads()
 
     /**
-     * @return the number of available reserved threads
+     * @return the number of available reserved threads that are currently reserved
```
The word `currently` could be assumed implicitly given its natural meaning in a typical sentence, i.e. in this case there is probably some different meaning that could be phrased more explicitly. The description inside `@ManagedAttribute` also hasn't been updated.

gregw committed:
Merge branch 'jetty-12.1.x' into fix/jetty-12.1.x/13187/qtpReluctantReservation

gregw commented:
I've moved the javadoc changes to #13212 so they can get merged quickly.

gregw committed:
QueuedThreadPool will not reserve a thread with jobs waiting

Fix #13187 by:
 + Modified QTP's ReservedThreadExecutor so that it will not reserve a thread if there are queued jobs waiting.
 + Deprecating many meaningless getters that combined the concept of available for execute(task) with available for tryExecute(task).
 + Improved javadoc

gregw committed:
QueuedThreadPool will not reserve a thread with jobs waiting

Fix #13187 by:
 + Modified QTP's ReservedThreadExecutor so that it will not reserve a thread if there are queued jobs waiting.
 + Deprecating many meaningless getters that combined the concept of available for execute(task) with available for tryExecute(task).
 + Improved javadoc

gregw committed:
QueuedThreadPool will not reserve a thread with jobs waiting

Fix #13187 by:
 + Modified QTP's ReservedThreadExecutor so that it will not reserve a thread if there are queued jobs waiting.
 + Deprecating many meaningless getters that combined the concept of available for execute(task) with available for tryExecute(task).
 + Improved javadoc

gregw committed:
QueuedThreadPool will not reserve a thread with jobs waiting

Fix #13187 by:
 + Modified QTP's ReservedThreadExecutor so that it will not reserve a thread if there are queued jobs waiting.
 + Deprecating many meaningless getters that combined the concept of available for execute(task) with available for tryExecute(task).
 + Improved javadoc

gregw committed:
updates from review

gregw committed:
updates from review

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/jetty-12.1.x/13187/qtpReluctantReservation

# Conflicts:
#	jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/BlockedWritesWithSmallThreadPoolTest.java
#	jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

gregw committed:
Merge branch 'jetty-12.1.x' into fix/jetty-12.1.x/13187/qtpReluctantReservation

lorban committed:
add AES+QTP limit test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
I've written an attempt at a limit test to see how this change impacts performance, see the new `AESLimit` class.

On my machine, it seems that with this change:
 - the latency doubles up to the 99.9th percentile
 - then lowers by ~20% up to the 99.999th
 - then doubles again up to the max latency

but throughput increases by ~40%:
 - with the patch, `AdaptiveExecutionStrategy` reports pec=4_777_050,epc=2_376_119
 - without the patch `AdaptiveExecutionStrategy` reports pec=1_486_968,epc=3_178_115

Assuming the test I wrote is sane, it's clear that `AdaptiveExecutionStrategy` successfully trades throughput for latency and that the patch noticeably degrades that tradeoff. Since the test I wrote does almost nothing in the executed task, I can't say for sure if the extra throughput would be maintained when the executed task would suffer from cache misses due to the context change.

I'm not sure how deep we should dig in, but it seems apparent that if we value EPC, this change has a negative impact.

lorban committed:
update AES+QTP limit test to try to be more realistic

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
After discussing with @sbordet, we agreed that the `AESLimit` test was not measuring what it was supposed to. Another attempt had been made, and the fundamental end result is similar (even if the details are widely different): this change degrades latency overall.

We need to have an agreement about if the latest test is measuring the right things or not before moving on.

gregw committed:
Fixed tests

gregw committed:
checkstyle

gregw commented:
bumped to 12.1.1 at least

gregw commented:
@sbordet @lorban There are only a few explanations for this PR causing increased latency:
 1) The test/benchmark is wrong
 2) The testing infrastructure is not stable and it was happen stance
 3) The extra test to check if there are tasks waiting is expensive... but it is a lock free check, conducted only by a small minority of threads between jobs.    Although - I guess it would be worthwhile to somehow measure how many threads running once were reserved, so we know how many are going to do this check
 4) The reduction of available reserved threads is impacting performance.  Which would be OK, as it indicates they are useful!

I currently think it is likely to be 1 or 2.  But it would be interesting to measure 3.... if that can be done without adding additional latency :)    It probably can be done with an inspection of the stack traces of all threads?

lorban commented:
> @sbordet @lorban There are only a few explanations for this PR causing increased latency:
> 
>     1. The test/benchmark is wrong

It's always a possibility that the benchmark isn't measuring what we think it's measuring. Or that it's not representative of the reality.

>     2. The testing infrastructure is not stable and it was happen stance

Always a possibility too.

>     3. The extra test to check if there are tasks waiting is expensive... but it is a lock free check, conducted only by a small minority of threads between jobs.    Although - I guess it would be worthwhile to somehow measure how many threads running once were reserved, so we know how many are going to do this check

Here, the benchmark explicitly configures a QTP of 4 threads: 1 selector, 1 acceptor and 2 spares with a max reserved thread of 1 so that what's measured is what happens when the handling jobs are fighting against the reserved threads for one of the 2 spare threads.

Maybe we should not care too much about this constrained environment, but it was the original question that started this perf exercise.

>     4. The reduction of available reserved threads is impacting performance.  Which would be OK, as it indicates they are useful!

To be strictly precise: the reduction of available reserved threads is **negatively** impacting **latency** but _seems_ it may have a positive impact on throughput. But I agree that we should and do value latency over throughput so I assume you saw the result of this benchmark as a degradation (assuming the benchmark is right, of course).

> I currently think it is likely to be 1 or 2. But it would be interesting to measure 3.... if that can be done without adding additional latency :) It probably can be done with an inspection of the stack traces of all threads?

The benchmark can already do that on the cheap by adding the async profiler to the JMH profiler list. IIRC we did that and the impact wasn't related to any form of contention but rather on the fact that a reserved thread was less often available in the constrained environment of the benchmark.

gregw commented:
@lorban @sbordet I think that measuring throughput and latency on extremely constrained threadpools is interesting, but not what we are optimising for.   Such deployments are concerned about footprint and so long as the requests are handled in some reasonable fashion, I do not think either latency or throughput are critical.  This PR is not about making constrained environments run faster/better, it is about making them run at all.   Without this PR, there are scenarios where a request to a constrained environment will never be run in a timely manner and we can effectively live lock the server as a thread is busy being reserved when it could handle a queued task.       So the test in this PR checks for that scenarios.    

I don't think the absolute performance numbers of a constrained environment are critical.  However, it is significant that changing the number of reserved threads does affect the latency/throughput tradeoff... exactly as we expect it should if EWYK and reserved threads are doing what they were designed to do.   I think this is a good datapoint to say that these optimisations are still relevant.

The benchmarks that I think are key for this PR is with a non constrained QTP (i.e. our normal benchmarks).   I would expect that for typical usage (when we often do not max out reserved threads), then this change will have very little effect, as we run low on reserved threads because of this, one more can be allocated.   Thus in a non-constrained test, I would hope to see little to no difference with or without this PR.       If there is no performance impact on a non-constrained environment, then this PR is worthwhile to avoid the live-lock in constrained environments

gregw committed:
Merge branch 'jetty-12.1.x' into fix/jetty-12.1.x/13187/qtpReluctantReservation

gregw commented:
@lorban I think your benchmark in this PR is testing the wrong thing.    

When running with a constrained ThreadPool, the issue is not about throughput or even latency.  It is about correctness.   Without this PR, there are some situations that will just hang forever as a thread is reserved instead of being used to run queued jobs.  With this PR, those scenarios do not hang and the unit tests I added pass.

The latency/throughput issues that we are worried about are only at our normal tested scale with our normal number of threads available.  We just need to know if the extra checks for correctness and potentially fewer reserved threads are costing us anything when we are not constrained.

wendigo commented:
I think that I'm observing this behavior in the Trino under load with HTTP/2. When it happens, http server "hangs" until some of the requests times out, then it continues to work. This happens from time to time and eventually server is able to recover.

gregw committed:
Merge branch 'jetty-12.1.x' into fix/jetty-12.1.x/13187/qtpReluctantReservation

gregw committed:
cleanups

gregw commented:
Thanks @wendigo.  
@sbordet @lorban I think this reinforces that this is a PR about correctness rather than performance.  We should not reserve threads if there are jobs available and no available other threads to run them.

gregw committed:
simplified algorithm

gregw committed:
fixed bad logic inversion (thanks IDE!)

gregw committed:
fixed bad logic inversion (thanks IDE!)

lorban committed:
remove System.err.println() calls

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
This has no measurable perf impact in the Core API: https://jenkins.webtide.net/job/load_testing/job/jetty-profiler-12.1.x/44/artifact/core/target/reports/CoreHandlerPerfTest_testNoGzipAsync/ according to our standard benchmark once I removed the `System.err.println()` calls in `isReservable()`.

(main branch reference: https://jenkins.webtide.net/job/load_testing/job/jetty-profiler-12.1.x/41/artifact/core/target/reports/CoreHandlerPerfTest_testNoGzipFullyAsyncHandlerTree/)

lorban commented:
Reminder of our last discussion: I'm okay with this change (after its gets cleaned up and reviewed) if we also change the heuristic that picks the default # of reserved thread to also do _if QTP.maxSize() < N then reservedThreads = 0_, with N still being open for discussion.

gregw committed:
Merge branch 'jetty-12.1.x' into fix/jetty-12.1.x/13187/qtpReluctantReservation

gregw committed:
No reserved Threads when low threads

gregw committed:
No reserved Threads when low threads

gregw committed:
No reserved Threads when low threads

gregw committed:
No reserved Threads when low threads

gregw commented:
@lorban I changed the heuristic with N=#cores

lorban commented:
@gregw I think we should use a hardcoded number like 8 or 10 instead of the #cores as running Jetty on some 2 vCores K8S virtual machines is far from uncommon.

FYI, I've been perf testing this branch and as usual, we have to discuss the results of the benchmark.

gregw committed:
updates from review

gregw commented:
@lorban I hard coded to 8.  Let's chat in your morning.

gregw commented:
On discussion, it was decided that this is not really a problem we should fix.   In essence the reserved thread is just a blocking job in the queue ahead of other jobs.  The fact that it is our job doesn't mean it should give up doing its job.
We can revisit at some future point.

scscgit commented:
@gregw Just curious: is there also no plan to mention this bug in the doc+javadoc, so that newcomers never use this setup until it's fixed? (I also imagine that one more thread could simply be allocated by default in addition to what the user configures.) Thanks

wendigo commented:
Yeah, documenting this would be awesome

gregw commented:
It is hard to describe this as a bug.

If you configure a thread pool with 5 threads and execute 6 blocking tasks, then the 6th of them will be deferred until one of the other tasks complete.
It just so happens that jetty does execute a blocking task - a reserved thread task.  So with this PR, if the reserved thread task is the 5th task to be executed AND the 6th task is executed before the reserved thread task begins, then the reserved thread can choose not to block.    But it is a pretty weak protection.  The 5th task is much more likely to be an application blocking task.

We might look at reducing the reserved thread timeout if there is a task queue, but ultimately if you have N threads in your thread pool and O(N) blocking tasks, then avoiding a reserved thread is only going to help you a little bit.  If you care, configure 0 reserved threads.

We are working through some related configuration at the moment and will improve documentation when we are done.

scscgit commented:
Well, the bugs I'm aware of are namely the lack of a FIFO guarantee and falsely ignoring `qosHandler.setMaxRequestCount(1)`. It's just about being explicit in the specification by describing the current behavior/contracts. I also believe that users may not be aware of any implications of "avoiding a reserved thread" from a performance standpoint, etc. Just make sure developers don't spend a hundred hours debugging such a pointless problem next time please, especially when we write integration tests and fail sanity check assertions :)
